src/ore/src/process.rs

```diff
@@ -0,0 +1,45 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Process utilities.
+
+/// Halts the process.
+///
+/// `halt` forwards the provided arguments to the [`tracing::warn`] macro, then
+/// terminates the process with exit code 2.
+///
+/// Halting a process is a middle ground between a graceful shutdown and a
+/// panic. Use halts for errors that are severe enough to require shutting down
+/// the process, but not severe enough to be considered a crash. Halts are not
+/// intended to be reported to error tracking tools (e.g., Sentry).
+///
+/// There are two common classes of errors that trigger a halt:
+///
+///   * An error that indicates a new process has taken over (e.g., an error
+///     indicating that the process's leader epoch has expired).
+///
+///   * A retriable error where the retry logic is too difficult to write
+///     correctly. Sometimes the easiest way to retry an error is to tear down
+///     the process and allow a new process to start from scratch.
+///
+#[cfg_attr(nightly_doc_features, doc(cfg(feature = "tracing")))]
+#[cfg(feature = "tracing")]
+#[macro_export]
+macro_rules! halt {
+    ($($arg:expr),* $(,)?) => {{
+        $crate::__private::tracing::warn!("halting process: {}", format!($($arg),*));
```

Comment:
in the interest of not regressing, can we capture a `Backtrace` (https://docs.rs/backtrace/latest/backtrace/struct.Backtrace.html#method.new) and (debug) print it here?

Reply:
Ideally we would also `tracing::error` with a `mz_halt` target as well, and use the new filtering to have one spot to control whether or not we want the error logs in production